apps/unicop/blog/2022-07-26-js-module-systems-series/index.md

```diff
@@ -0,0 +1,80 @@
+
+🧩 **[Asynchronous Module Definition, or shortly "AMD"](https://jameshfisher.com/2020/10/03/what-are-amd-modules/)**, where a module calls `define(dependencies, callback)`.
+
+- module keywords: `define`
+
+🧩 **[CommonJS](https://jameshfisher.com/2020/09/27/what-does-the-require-function-do-in-nodejs/)**, where a module’s dependencies are synchronously, dynamically `require()`d, and its exports are placed on an `module.exports object`.
+
+- module keywords: `require`, `module.exports`, `exports`
+
+🧩 **[“ECMAScript modules”](https://jameshfisher.com/2020/09/25/javascript-modules-for-grumpy-developers-from-2005/)**, or `"ESM"`, where a module’s dependencies are statically imported before execution, and its exports are statically defined, as top-level variables.
+
+- module keywords: `import`, `export`, `export default`
+
+As you can see, the eco-system is messy, but the mess just drove it to develop a fifth module system (😵‍💫), that makes order in the mess, and that is `Universal Module Definition`, or shortly `“UMD”`.
+
+🧩 **[UMD module](https://jameshfisher.com/2020/10/04/what-are-umd-modules/)** is a `JavaScript` script, that guess during runtime which module system it’s being used in, and then it acts as that kind of module. So you can load the file in a plain `<script>`, or you can load it from an `AMD` module loader, or you can load it as a `Node.js module`, and it will always do something sensible. Here is the **[basic pattern](https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js)** that makes `UMD` to work.
+
+- Important Note: `UMD` supports all module systems but `ESM`.
+
+###### Credit 🎖️
+
+Inspired by **[jameshfisher](https://jameshfisher.com/)**
+IMO he has great materials.
+
+## What JavaScript based technology runs what module systems
+
+✅ - Natively | 🧪 - Experimental | 📦 - With Module Bundler (e.g. webpack)
+
+| Module System  | Browsers | Nodejs | <sub><sup>Nodejs based [1]</sup></sub> CLIs |
+| :------------: | :------: | :----: | :-----------------------------------------: |
+| Raw `<script>` |    ✅    |        |                                             |
+|      AMD       |    ✅    |        |                                             |
+|    CommonJS    |    📦    |   ✅   |                     ✅                      |
+|      UMD       |    ✅    |   ✅   |                     ✅                      |
+|      ESM       |    ✅    |   🧪   |                     🧪                      |
+
+<sub><sup>[1] - <a href="https://jestjs.io/"><b>jest</b></a> for instance is a Nodejs based CLI</sup></sub>
+
+## The isomorphic code confusion 🔮
+
+The less obvious complication is browser compatibility.
+
+Both Browsers and The NodeJS ecosystems initially started separate, but since tools like `webpack`, `parcels`, `rollup` and others enabled CommonJS modules to be used in the browser, the scene of isomorphic code, which runs both under NodeJS and the browser, has exploded.
```

Comment:
```suggestion
Both Browsers and The NodeJS ecosystems initially started separately, but since tools like `webpack`, `parcels`, `rollup` and others enabled CommonJS modules to be used in the browser, the scene of isomorphic code, which runs both under NodeJS and the browser, has exploded.
```